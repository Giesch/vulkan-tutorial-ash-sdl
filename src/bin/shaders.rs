use std::process::Command;

const SHADERS_SOURCE_DIR: &'static str = "./shaders/source";
const SHADERS_COMPILED_DIR: &'static str = "./shaders/compiled";

/// Compiles each glsl source shader into spv
/// Depends on an installed glslc (from the vulkan sdk)
pub fn main() {
    let shader_source_dir = std::fs::read_dir(SHADERS_SOURCE_DIR).unwrap();

    if !std::fs::exists(SHADERS_COMPILED_DIR).unwrap() {
        std::fs::create_dir(SHADERS_COMPILED_DIR).unwrap();
    }

    for entry in shader_source_dir {
        let entry = entry.unwrap();

        let in_path = entry.path().display().to_string();

        let file_name = entry.file_name();
        let file_name = file_name.to_string_lossy();

        let out_file_name = file_name.replace("glsl", "spv");
        let out_path = format!("{SHADERS_COMPILED_DIR}/{out_file_name}");

        let shader_stage = if file_name.contains(".vert") {
            "vert"
        } else if file_name.contains(".frag") {
            "frag"
        } else {
            panic!("unable to determine shader stage for {}", file_name);
        };

        let output = Command::new("glslc")
            .arg(format!("-fshader-stage={shader_stage}"))
            .arg(&in_path)
            .arg("-o")
            .arg(&out_path)
            .output()
            .unwrap();

        if !output.status.success() {
            dbg!(&output);
            panic!("failed to compile shader");
        }
    }
}
