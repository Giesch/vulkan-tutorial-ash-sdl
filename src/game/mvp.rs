use crate::renderer::gpu_write::GPUWrite;

#[derive(Debug, Clone)]
#[repr(C, align(16))]
pub struct MVPMatrices {
    pub model: glam::Mat4,
    pub view: glam::Mat4,
    pub projection: glam::Mat4,
}

impl GPUWrite for MVPMatrices {}
