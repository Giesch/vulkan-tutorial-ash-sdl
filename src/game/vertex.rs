use crate::renderer::gpu_write::GPUWrite;

#[derive(Debug, Clone)]
#[repr(C, align(16))]
pub struct Vertex {
    pub position: glam::Vec3,
    pub color: glam::Vec3,
    pub tex_coord: glam::Vec2,
}

impl GPUWrite for Vertex {}

impl Vertex {
    pub fn binding_description() -> ash::vk::VertexInputBindingDescription {
        ash::vk::VertexInputBindingDescription::default()
            .binding(0)
            .stride(std::mem::size_of::<Self>() as u32)
            .input_rate(ash::vk::VertexInputRate::VERTEX)
    }

    pub fn attribute_descriptions() -> [ash::vk::VertexInputAttributeDescription; 3] {
        // color formats are also used to define non-color vec sizes 1-4
        //   (the official tutorial is mildly apologetic)
        // BUT this does matter for defaults -
        //   if there aren't enough components here to fill the components shader-side,
        //   the 'color' components default to 0 and 'alpha' component defaults to 1
        let glam_vec_3_format = ash::vk::Format::R32G32B32_SFLOAT;
        let glam_vec_2_format = ash::vk::Format::R32G32_SFLOAT;

        // the bindings and locations here are about how this data is read
        // out of the vertex buffer, not the binding to the shader
        // bindings - the index in the array passed to cmd_bind_vertex_buffers
        //   in our case, always 0 because there is only one
        // locations - a unique identifier for the attribute
        [
            // position
            ash::vk::VertexInputAttributeDescription::default()
                .binding(0)
                .location(0)
                .format(glam_vec_3_format)
                .offset(std::mem::offset_of!(Vertex, position) as u32),
            // color
            ash::vk::VertexInputAttributeDescription::default()
                .binding(0)
                .location(1)
                .format(glam_vec_3_format)
                .offset(std::mem::offset_of!(Vertex, color) as u32),
            // texture coordinate
            ash::vk::VertexInputAttributeDescription::default()
                .binding(0)
                .location(2)
                .format(glam_vec_2_format)
                .offset(std::mem::offset_of!(Vertex, tex_coord) as u32),
        ]
    }
}
