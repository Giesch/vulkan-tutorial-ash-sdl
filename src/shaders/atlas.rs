use std::ffi::CString;

use ash::vk;

mod depth_texture;
pub use depth_texture::*;

use crate::renderer::LayoutDescription;

use super::json::ReflectedPipelineLayout;

pub struct ShaderAtlas {
    pub depth_texture: DepthTextureShader,
}

impl ShaderAtlas {
    pub fn init() -> Self {
        Self {
            depth_texture: DepthTextureShader::init(),
        }
    }
}

pub trait ShaderAtlasEntry {
    // dev only

    // used in hot reload
    fn source_file_name(&self) -> &str;

    // dev and release

    fn uniform_buffer_sizes(&self) -> Vec<u64>;

    fn vertex_binding_descriptions(&self) -> Vec<vk::VertexInputBindingDescription>;
    fn vertex_attribute_descriptions(&self) -> Vec<vk::VertexInputAttributeDescription>;

    // one set of descriptions per descriptor set
    fn layout_bindings(&self) -> Vec<Vec<LayoutDescription>>;

    // release only

    fn precompiled_shaders(&self) -> PrecompiledShaders;

    fn pipeline_layout(&self) -> &ReflectedPipelineLayout;
}

pub struct PrecompiledShaders {
    pub vert: PrecompiledShader,
    pub frag: PrecompiledShader,
}

pub struct PrecompiledShader {
    pub entry_point_name: CString,
    pub spv_bytes: Vec<u32>,
}
