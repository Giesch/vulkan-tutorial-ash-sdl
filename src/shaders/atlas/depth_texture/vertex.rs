use crate::renderer::vertex_description::VertexDescription;

pub use crate::generated::shader_atlas::depth_texture::Vertex;

impl VertexDescription for Vertex {
    fn binding_descriptions() -> Vec<ash::vk::VertexInputBindingDescription> {
        let binding_description = ash::vk::VertexInputBindingDescription::default()
            .binding(0)
            .stride(std::mem::size_of::<Self>() as u32)
            .input_rate(ash::vk::VertexInputRate::VERTEX);

        vec![binding_description]
    }

    fn attribute_descriptions() -> Vec<ash::vk::VertexInputAttributeDescription> {
        // color formats are also used to define non-color vec sizes 1-4
        //   (the official tutorial is mildly apologetic)
        // BUT this does matter for defaults -
        //   if there aren't enough components here to fill the components shader-side,
        //   the 'color' components default to 0 and 'alpha' component defaults to 1
        let glam_vec_3_format = ash::vk::Format::R32G32B32_SFLOAT;
        let glam_vec_2_format = ash::vk::Format::R32G32_SFLOAT;

        // the bindings and locations here are about how this data is read
        // out of the vertex buffer, not the binding to the shader
        // bindings - the index in the array passed to cmd_bind_vertex_buffers
        //   in our case, always 0 because there is only one
        // locations - a unique identifier for the attribute
        vec![
            // position
            ash::vk::VertexInputAttributeDescription::default()
                .binding(0)
                .location(0)
                .format(glam_vec_3_format)
                .offset(std::mem::offset_of!(Vertex, position) as u32),
            // color
            ash::vk::VertexInputAttributeDescription::default()
                .binding(0)
                .location(1)
                .format(glam_vec_3_format)
                .offset(std::mem::offset_of!(Vertex, color) as u32),
            // texture coordinate
            ash::vk::VertexInputAttributeDescription::default()
                .binding(0)
                .location(2)
                .format(glam_vec_2_format)
                .offset(std::mem::offset_of!(Vertex, tex_coord) as u32),
        ]
    }
}
