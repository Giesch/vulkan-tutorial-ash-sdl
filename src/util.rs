use std::path::PathBuf;

pub fn manifest_path<'a>(segments: impl IntoIterator<Item = &'a str>) -> PathBuf {
    let segments = segments.into_iter();
    let full_path = [env!("CARGO_MANIFEST_DIR")].into_iter().chain(segments);
    full_path.collect()
}
