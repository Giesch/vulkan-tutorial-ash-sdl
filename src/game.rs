use std::ffi::c_void;
use std::path::PathBuf;
use std::time::Instant;

use anyhow::Context;
use image::{DynamicImage, ImageReader};

use crate::util::manifest_path;

use super::shaders::COLUMN_MAJOR;

mod vertex;
use vertex::*;

mod mvp;
use mvp::MVPMatrices;

// TODO
// we want a more zoomed-out api than this
//
// like, Game should say,
// 'I want these GPU resources: a uniform buffer of size x, ...', etc
//
// then the renderer returns typed indirect handles that persist through hot reload
//   so the game can keep those handles in its own struct
//   someday maybe give them a 'scene' lifetime? not worth it?
//
// then, in an 'update' callback on game, called by app
//   game calls renderer with the handles to get access to pointers for the frame
//   or renderer has higher-level methods that take a handle
//
// game makes some kind of create pipeline request to the renderer
//   gets a pipeline object handle back, or a struct w/a set of handles?
//   or the renderer has an internal 'pipeline description' type created that it uses
// can we get that whole request/description from reflection?
//   not entirely; need to be able to load textures and such
//   so a combination of the reflected shader-atlas entry and other resources
// how can we map the create request struct into the created resources struct?
//   need a macro
pub trait Game {
    fn uniform_buffer_size(&self) -> usize;

    fn update_uniform_buffer(
        &self,
        aspect_ratio: f32,
        mapped_uniform_buffer: *mut c_void,
    ) -> anyhow::Result<()>;

    fn load_vertices(&self) -> Result<(Vec<Vertex>, Vec<u32>), anyhow::Error>;
    fn load_texture(&self) -> Result<DynamicImage, anyhow::Error>;

    fn vertex_binding_descriptions(&self) -> Vec<ash::vk::VertexInputBindingDescription>;
    fn vertex_attribute_descriptions(&self) -> Vec<ash::vk::VertexInputAttributeDescription>;
}

#[allow(unused)]
pub struct VikingRoom {
    start_time: Instant,
}

#[allow(unused)]
impl VikingRoom {
    pub fn init() -> Self {
        let start_time = Instant::now();
        Self { start_time }
    }
}

impl Game for VikingRoom {
    fn load_texture(&self) -> Result<DynamicImage, anyhow::Error> {
        load_image("viking_room.png")
    }

    fn uniform_buffer_size(&self) -> usize {
        std::mem::size_of::<MVPMatrices>()
    }

    fn update_uniform_buffer(
        &self,
        aspect_ratio: f32,
        mapped_uniform_buffer: *mut c_void,
    ) -> anyhow::Result<()> {
        update_mvp_uniform_buffer(
            self.start_time,
            aspect_ratio,
            mapped_uniform_buffer as *mut MVPMatrices,
        )
    }

    fn vertex_binding_descriptions(&self) -> Vec<ash::vk::VertexInputBindingDescription> {
        vec![Vertex::binding_description()]
    }

    fn vertex_attribute_descriptions(&self) -> Vec<ash::vk::VertexInputAttributeDescription> {
        Vertex::attribute_descriptions().to_vec()
    }

    // From unknownue's rust version
    // https://github.com/unknownue/vulkan-tutorial-rust/blob/master/src/tutorials/27_model_loading.rs
    fn load_vertices(&self) -> Result<(Vec<Vertex>, Vec<u32>), anyhow::Error> {
        let file_path: PathBuf = [env!("CARGO_MANIFEST_DIR"), "models", "viking_room.obj"]
            .iter()
            .collect();

        let (mut models, _materials) = tobj::load_obj(file_path, &tobj::GPU_LOAD_OPTIONS)?;

        debug_assert!(models.len() == 1);
        let model = models.remove(0);

        let mut vertices = vec![];
        let mesh = model.mesh;
        let vertices_count = mesh.positions.len() / 3;
        for i in 0..vertices_count {
            let position = {
                let offset = i * 3;
                glam::Vec3::new(
                    mesh.positions[offset],
                    mesh.positions[offset + 1],
                    mesh.positions[offset + 2],
                )
            };

            let tex_coord = {
                let offset = i * 2;
                let u = mesh.texcoords[offset];
                // in obj, 0 is the bottom, in vulkan, 0 is the top
                // (for texture coordinates)
                let v = 1.0 - mesh.texcoords[offset + 1];
                glam::Vec2::new(u, v)
            };

            let vertex = Vertex {
                position,
                color: glam::Vec3::splat(1.0),
                tex_coord,
            };

            vertices.push(vertex);
        }

        Ok((vertices, mesh.indices))
    }
}

fn update_mvp_uniform_buffer(
    start_time: Instant,
    aspect_ratio: f32,
    mapped_uniform_buffer: *mut MVPMatrices,
) -> Result<(), anyhow::Error> {
    const TURN_DEGREES_PER_SECOND: f32 = 5.0;
    const STARTING_ANGLE_DEGREES: f32 = 45.0;

    let elapsed_seconds = (Instant::now() - start_time).as_secs_f32();
    let turn_radians = elapsed_seconds * TURN_DEGREES_PER_SECOND.to_radians();

    let model = glam::Mat4::from_rotation_z(turn_radians);
    let view = glam::Mat4::look_at_rh(
        glam::Vec3::splat(2.0),
        glam::Vec3::splat(0.0),
        glam::Vec3::new(0.0, 0.0, 1.0),
    );
    let projection =
        glam::Mat4::perspective_rh(STARTING_ANGLE_DEGREES.to_radians(), aspect_ratio, 0.1, 10.0);

    let mut mvp = MVPMatrices {
        model,
        view,
        projection,
    };

    // "GLM was originally designed for OpenGL,
    // where the Y coordinate of the clip coordinates is inverted.
    // The easiest way to compensate for that is to flip the sign
    // on the scaling factor of the Y axis in the projection matrix.
    // If you donâ€™t do this, then the image will be rendered upside down."
    // https://docs.vulkan.org/tutorial/latest/05_Uniform_buffers/00_Descriptor_set_layout_and_buffer.html
    mvp.projection.y_axis.y *= -1.0;

    if !COLUMN_MAJOR {
        // it's also possible to avoid this by reversing the mul() calls in shaders
        // https://discord.com/channels/1303735196696445038/1395879559827816458/1396913440584634499
        mvp.model = mvp.model.transpose();
        mvp.view = mvp.view.transpose();
        mvp.projection = mvp.projection.transpose();
    }

    unsafe {
        std::ptr::copy_nonoverlapping(&mvp, mapped_uniform_buffer as *mut MVPMatrices, 1);
    }

    Ok(())
}

#[allow(unused)]
pub struct DepthTexture {
    start_time: Instant,
}

#[allow(unused)]
impl DepthTexture {
    pub fn init() -> Self {
        let start_time = Instant::now();
        Self { start_time }
    }
}

impl Game for DepthTexture {
    fn load_texture(&self) -> Result<DynamicImage, anyhow::Error> {
        load_image("texture.jpg")
    }

    fn uniform_buffer_size(&self) -> usize {
        std::mem::size_of::<MVPMatrices>()
    }

    fn update_uniform_buffer(
        &self,
        aspect_ratio: f32,
        mapped_uniform_buffer: *mut c_void,
    ) -> anyhow::Result<()> {
        update_mvp_uniform_buffer(
            self.start_time,
            aspect_ratio,
            mapped_uniform_buffer as *mut MVPMatrices,
        )
    }

    fn vertex_binding_descriptions(&self) -> Vec<ash::vk::VertexInputBindingDescription> {
        vec![Vertex::binding_description()]
    }

    fn vertex_attribute_descriptions(&self) -> Vec<ash::vk::VertexInputAttributeDescription> {
        Vertex::attribute_descriptions().to_vec()
    }

    fn load_vertices(&self) -> Result<(Vec<Vertex>, Vec<u32>), anyhow::Error> {
        let vertices = vec![
            Vertex {
                position: glam::Vec3::new(-0.5, -0.5, 0.0),
                color: glam::Vec3::new(1.0, 0.0, 0.0),
                tex_coord: glam::Vec2::new(1.0, 0.0),
            },
            Vertex {
                position: glam::Vec3::new(0.5, -0.5, 0.0),
                color: glam::Vec3::new(0.0, 1.0, 0.0),
                tex_coord: glam::Vec2::new(0.0, 0.0),
            },
            Vertex {
                position: glam::Vec3::new(0.5, 0.5, 0.0),
                color: glam::Vec3::new(0.0, 0.0, 1.0),
                tex_coord: glam::Vec2::new(0.0, 1.0),
            },
            Vertex {
                position: glam::Vec3::new(-0.5, 0.5, 0.0),
                color: glam::Vec3::new(1.0, 1.0, 1.0),
                tex_coord: glam::Vec2::new(1.0, 1.0),
            },
            Vertex {
                position: glam::Vec3::new(-0.5, -0.5, -0.5),
                color: glam::Vec3::new(1.0, 0.0, 0.0),
                tex_coord: glam::Vec2::new(1.0, 0.0),
            },
            Vertex {
                position: glam::Vec3::new(0.5, -0.5, -0.5),
                color: glam::Vec3::new(0.0, 1.0, 0.0),
                tex_coord: glam::Vec2::new(0.0, 0.0),
            },
            Vertex {
                position: glam::Vec3::new(0.5, 0.5, -0.5),
                color: glam::Vec3::new(0.0, 0.0, 1.0),
                tex_coord: glam::Vec2::new(0.0, 1.0),
            },
            Vertex {
                position: glam::Vec3::new(-0.5, 0.5, -0.5),
                color: glam::Vec3::new(1.0, 1.0, 1.0),
                tex_coord: glam::Vec2::new(1.0, 1.0),
            },
        ];

        #[rustfmt::skip]
        let indices = vec![
            0, 1, 2, 2, 3, 0,
            4, 5, 6, 6, 7, 4,
        ];

        Ok((vertices, indices))
    }
}

fn load_image(file_name: &str) -> anyhow::Result<DynamicImage> {
    let file_path = manifest_path(["textures", file_name]);
    let image = ImageReader::open(&file_path)
        .with_context(|| format!("failed to open image: {file_path:?}"))?
        .decode()
        .with_context(|| format!("failed to decode image: {file_path:?}"))?;

    Ok(image)
}
