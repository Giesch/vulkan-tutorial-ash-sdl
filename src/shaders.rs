use std::ffi::CString;

use shader_slang as slang;
use shader_slang::Downcast;

use ash::vk::{self, Handle};

use crate::util::*;

/// whether to use column-major or row-major matricies with slang
pub const COLUMN_MAJOR: bool = true;

pub fn precompiled_shaders() -> Result<CompiledShaderModule, BoxError> {
    // TODO glob for all .slang/.spv files
    let spv_file_name = "depth_texture.spv";
    let path = manifest_path(["shaders", "compiled", spv_file_name]);

    let shader_bytecode = std::fs::read(&path).unwrap();
    let byte_reader = &mut std::io::Cursor::new(shader_bytecode.as_slice());
    let _spv_bytes = ash::util::read_spv(byte_reader)?;

    todo!()
}

pub fn compile_slang_shaders() -> Result<CompiledShaderModule, BoxError> {
    let global_session = slang::GlobalSession::new().unwrap();
    let search_path = CString::new("shaders/source")?;

    let session_options = slang::CompilerOptions::default()
        .vulkan_use_entry_point_name(true)
        .language(slang::SourceLanguage::Slang)
        .optimization(slang::OptimizationLevel::High)
        .emit_spirv_directly(true);
    let session_options = if COLUMN_MAJOR {
        session_options.matrix_layout_column(true)
    } else {
        session_options.matrix_layout_row(true)
    };

    let target_desc = slang::TargetDesc::default()
        .format(slang::CompileTarget::Spirv)
        .profile(global_session.find_profile("glsl_450+spirv_1_6"));

    let targets = [target_desc];
    let search_paths = [search_path.as_ptr()];
    let session_desc = slang::SessionDesc::default()
        .targets(&targets)
        .search_paths(&search_paths)
        .options(&session_options);

    let session = global_session.create_session(&session_desc).unwrap();
    // TODO glob for all .slang files
    let source_file_name = "depth_texture.slang";
    let module = session.load_module(source_file_name)?;

    // the examples have 1 vert and 1 frag shader
    debug_assert!(module.entry_points().len() == 2);

    let mut vert: Option<CompiledShader> = None;
    let mut frag: Option<CompiledShader> = None;
    for entry_point in module.entry_points() {
        let compiled_shader = compile_shader(&entry_point, &session, &module)?;

        if compiled_shader.stage == slang::Stage::Vertex {
            vert = Some(compiled_shader)
        } else if compiled_shader.stage == slang::Stage::Fragment {
            frag = Some(compiled_shader)
        }
    }

    match (vert, frag) {
        (Some(vertex_shader), Some(fragment_shader)) => Ok(CompiledShaderModule {
            source_file_name: source_file_name.into(),
            vertex_shader,
            fragment_shader,
        }),

        _ => Err(format!(
            "failed to load vertex and/or fragment entry points for: {source_file_name}"
        )
        .into()),
    }
}

#[derive(Debug)]
pub struct CompiledShaderModule {
    pub source_file_name: String,
    pub vertex_shader: CompiledShader,
    pub fragment_shader: CompiledShader,
}

// TODO add reflection metadata needed by vulkan
// need to generate the Vertex struct and its methods or const values
pub struct CompiledShader {
    pub entry_point_name: CString,
    pub stage: slang::Stage,
    pub spv_bytes: Vec<u32>,
}

impl std::fmt::Debug for CompiledShader {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("CompiledShader")
            .field("entry_point_name", &self.entry_point_name)
            .field("stage", &self.stage)
            .field("spv_bytes.len()", &self.spv_bytes.len())
            .finish()
    }
}

fn compile_shader(
    entry_point: &slang::EntryPoint,
    session: &slang::Session,
    module: &slang::Module,
) -> Result<CompiledShader, BoxError> {
    let program = session.create_composite_component_type(&[
        module.downcast().clone(),
        entry_point.downcast().clone(),
    ])?;

    let linked_program = program.link()?;

    let program_layout = linked_program.layout(0)?;

    let mut refl_entry_points = program_layout.entry_points();
    assert!(refl_entry_points.len() == 1);
    let reflection_entry_point = refl_entry_points.next().unwrap();
    let stage = reflection_entry_point.stage();

    let shader_bytecode: slang::Blob = linked_program.entry_point_code(0, 0)?;
    let byte_reader = &mut std::io::Cursor::new(shader_bytecode.as_slice());
    let spv_bytes = ash::util::read_spv(byte_reader)?;

    let entry_point_name = CString::new(reflection_entry_point.name())?;

    Ok(CompiledShader {
        entry_point_name,
        stage,
        spv_bytes,
    })
}

// slang reflection based vulkan builders
// https://docs.shader-slang.org/en/latest/parameter-blocks.html#using-parameter-blocks-with-reflection

pub struct PipelineLayoutBuilder {
    device: ash::Device,
    descriptor_set_layouts: Vec<vk::DescriptorSetLayout>,
    push_constant_ranges: Vec<vk::PushConstantRange>,
}

impl PipelineLayoutBuilder {
    pub fn new(device: ash::Device) -> Self {
        Self {
            device,
            descriptor_set_layouts: vec![],
            push_constant_ranges: vec![],
        }
    }

    pub fn add_push_constatant_range_for_constant_buffer(
        &mut self,
        constant_buffer_type_layout: &slang::reflection::TypeLayout,
    ) {
        let element_type_layout = constant_buffer_type_layout.element_type_layout();
        let element_size = element_type_layout.size(slang::ParameterCategory::Uniform);

        if element_size == 0 {
            return;
        }

        // NOTE this relies on the way the slang compiler
        // only ever uses one push constant range per entry point
        let offset = 0;

        let vk_push_constant_range = vk::PushConstantRange::default()
            // TODO use correct stage flags
            .stage_flags(vk::ShaderStageFlags::ALL)
            .offset(offset)
            .size(element_size as u32);

        self.push_constant_ranges.push(vk_push_constant_range);
    }

    fn add_sub_object_ranges(
        &mut self,
        type_layout: &slang::reflection::TypeLayout,
    ) -> Result<(), BoxError> {
        for sub_object_range_index in 0..type_layout.sub_object_range_count() {
            self.add_sub_object_range(type_layout, sub_object_range_index)?;
        }

        Ok(())
    }

    fn add_sub_object_range(
        &mut self,
        type_layout: &slang::reflection::TypeLayout,
        sub_object_range_index: i64,
    ) -> Result<(), BoxError> {
        let binding_range_index =
            type_layout.sub_object_range_binding_range_index(sub_object_range_index);
        let binding_type = type_layout.binding_range_type(binding_range_index);

        use slang::BindingType;

        match binding_type {
            BindingType::ParameterBlock => {
                let parameter_block_type_layout =
                    type_layout.binding_range_leaf_type_layout(binding_range_index);
                self.add_descriptor_set_for_parameter_block(parameter_block_type_layout)?;
            }

            BindingType::PushConstant => {
                let constant_buffer_type_layout =
                    type_layout.binding_range_leaf_type_layout(binding_range_index);
                self.add_push_constatant_range_for_constant_buffer(constant_buffer_type_layout);
            }

            // BindingType::Unknown => todo!(),
            // BindingType::Sampler => todo!(),
            // BindingType::Texture => todo!(),
            // BindingType::ConstantBuffer => todo!(),
            // BindingType::TypedBuffer => todo!(),
            // BindingType::RawBuffer => todo!(),
            // BindingType::CombinedTextureSampler => todo!(),
            // BindingType::InputRenderTarget => todo!(),
            // BindingType::InlineUniformData => todo!(),
            // BindingType::RayTracingAccelerationStructure => todo!(),
            // BindingType::VaryingInput => todo!(),
            // BindingType::VaryingOutput => todo!(),
            // BindingType::ExistentialValue => todo!(),
            // BindingType::MutableFlag => todo!(),
            // BindingType::MutableTeture => todo!(),
            // BindingType::MutableTypedBuffer => todo!(),
            // BindingType::MutableRawBuffer => todo!(),
            // BindingType::BaseMask => todo!(),
            // BindingType::ExtMask => todo!(),
            _ => {}
        }

        Ok(())
    }

    pub fn add_descriptor_set_for_parameter_block(
        &mut self,
        parameter_block_type_layout: &slang::reflection::TypeLayout,
    ) -> Result<(), BoxError> {
        let mut descriptor_set_layout_builder = DescriptorSetLayoutBuilder::default();
        descriptor_set_layout_builder.add_descriptor_ranges_for_parameter_block_element(
            parameter_block_type_layout.element_type_layout(),
            self,
        )?;

        descriptor_set_layout_builder.build_and_add(self)?;

        Ok(())
    }

    // aka 'finishBuilding' in the docs
    pub fn build(&mut self) -> Result<vk::PipelineLayout, BoxError> {
        // a null here represents an unused reserved slot for a
        // ParameterBlock that ended up only containing other ParameterBlocks
        // https://docs.shader-slang.org/en/latest/parameter-blocks.html#empty-parameter-blocks
        self.descriptor_set_layouts.retain(|l| !l.is_null());

        let pipeline_layout_info = vk::PipelineLayoutCreateInfo::default()
            .set_layouts(&self.descriptor_set_layouts)
            .push_constant_ranges(&self.push_constant_ranges);

        let pipeline_layout = unsafe {
            self.device
                .create_pipeline_layout(&pipeline_layout_info, None)?
        };

        Ok(pipeline_layout)
    }
}

#[derive(Default)]
pub struct DescriptorSetLayoutBuilder<'a> {
    set_index: usize,
    binding_ranges: Vec<vk::DescriptorSetLayoutBinding<'a>>,
}

impl<'a> DescriptorSetLayoutBuilder<'a> {
    pub fn new(pipeline_layout_builder: &mut PipelineLayoutBuilder) -> Self {
        // reserve a layout slot to be filled in later
        // this preserves the correct index order for nested ParameterBlocks
        // https://docs.shader-slang.org/en/latest/parameter-blocks.html#ordering-of-nested-parameter-blocks

        let set_index = pipeline_layout_builder.descriptor_set_layouts.len();

        pipeline_layout_builder
            .descriptor_set_layouts
            .push(vk::DescriptorSetLayout::null());

        let binding_ranges = vec![];

        Self {
            set_index,
            binding_ranges,
        }
    }

    pub fn add_descriptor_ranges_for_parameter_block_element(
        &mut self,
        element_layout: &slang::reflection::TypeLayout,
        pipeline_layout_builder: &mut PipelineLayoutBuilder,
    ) -> Result<(), BoxError> {
        // in the cpp header there's a default argument overload for Uniform
        if element_layout.size(slang::ParameterCategory::Uniform) > 0 {
            self.add_automatically_introduced_uniform_buffer();
        }

        self.add_descriptor_ranges(element_layout);
        pipeline_layout_builder.add_sub_object_ranges(element_layout)?;

        Ok(())
    }

    fn add_automatically_introduced_uniform_buffer(&mut self) {
        // this relies on using no manual binding annotations
        let vk_binding_index = self.binding_ranges.len();

        let binding = vk::DescriptorSetLayoutBinding::default()
            .stage_flags(vk::ShaderStageFlags::ALL)
            .binding(vk_binding_index as u32)
            .descriptor_count(1)
            .descriptor_type(vk::DescriptorType::UNIFORM_BUFFER);

        self.binding_ranges.push(binding)
    }

    fn add_descriptor_ranges(&mut self, type_layout: &slang::reflection::TypeLayout) {
        // NOTE this means we are only querying the first descriptor set
        // doing this is vulkan-specific
        let relative_set_index = 0;

        let range_count = type_layout.descriptor_set_descriptor_range_count(relative_set_index);

        for range_index in 0..range_count {
            self.add_descriptor_range(type_layout, relative_set_index, range_index);
        }
    }

    fn add_descriptor_range(
        &mut self,
        type_layout: &slang::reflection::TypeLayout,
        relative_set_index: i64,
        range_index: i64,
    ) {
        let binding_type =
            type_layout.descriptor_set_descriptor_range_type(relative_set_index, range_index);
        if binding_type == slang::BindingType::PushConstant {
            // this is accounted for in add_sub_object_range
            // TODO should this also skip a nested ParameterBlock?
            return;
        }

        let descriptor_count = type_layout
            .descriptor_set_descriptor_range_descriptor_count(relative_set_index, range_index);

        // TODO what goes in the '...' here?
        // https://docs.shader-slang.org/en/latest/parameter-blocks.html#descriptor-ranges

        // this relies on using no manual binding annotations
        let vk_binding_index = self.binding_ranges.len();
        let descriptor_type = map_slang_binding_type_to_vk_descriptor_type(binding_type);

        let descriptor_set_layout_binding = vk::DescriptorSetLayoutBinding::default()
            .binding(vk_binding_index as u32)
            .descriptor_count(descriptor_count as u32)
            // TODO where to get these from? '_currentStageFlags' in the docs
            .stage_flags(vk::ShaderStageFlags::ALL)
            .descriptor_type(descriptor_type);

        self.binding_ranges.push(descriptor_set_layout_binding);
    }

    // aka 'finishBuilding' in the docs
    // creates a vulkan DescriptorSetLayout and adds it to the PipelineLayoutBuilder
    pub fn build_and_add(
        &self,
        pipeline_layout_builder: &mut PipelineLayoutBuilder,
    ) -> Result<(), BoxError> {
        if self.binding_ranges.is_empty() {
            return Ok(());
        }

        let create_info =
            vk::DescriptorSetLayoutCreateInfo::default().bindings(&self.binding_ranges);
        let layout = unsafe {
            pipeline_layout_builder
                .device
                .create_descriptor_set_layout(&create_info, None)?
        };

        pipeline_layout_builder.descriptor_set_layouts[self.set_index] = layout;

        Ok(())
    }
}

fn map_slang_binding_type_to_vk_descriptor_type(
    binding_type: slang::BindingType,
) -> vk::DescriptorType {
    use slang::BindingType;

    match binding_type {
        BindingType::Sampler => vk::DescriptorType::SAMPLER,
        BindingType::Texture => vk::DescriptorType::SAMPLED_IMAGE,

        BindingType::ConstantBuffer => vk::DescriptorType::UNIFORM_BUFFER,

        BindingType::ParameterBlock => todo!(),

        BindingType::VaryingInput => todo!(),
        BindingType::VaryingOutput => todo!(),
        BindingType::PushConstant => todo!(),
        BindingType::TypedBuffer => todo!(),
        BindingType::RawBuffer => todo!(),
        BindingType::CombinedTextureSampler => todo!(),
        BindingType::InputRenderTarget => todo!(),
        BindingType::InlineUniformData => todo!(),
        BindingType::RayTracingAccelerationStructure => todo!(),
        BindingType::ExistentialValue => todo!(),
        BindingType::MutableFlag => todo!(),
        BindingType::MutableTeture => todo!(),
        BindingType::MutableTypedBuffer => todo!(),
        BindingType::MutableRawBuffer => todo!(),
        BindingType::BaseMask => todo!(),
        BindingType::ExtMask => todo!(),
        BindingType::Unknown => todo!(),
    }
}

fn create_pipeline_layout(
    device: ash::Device,
    program_layout: &slang::reflection::Shader,
) -> Result<vk::PipelineLayout, BoxError> {
    let mut pipeline_layout_builder = PipelineLayoutBuilder::new(device);

    // this will hold top-level non-parameter-block shader parameters
    let default_descriptor_set_layout_builder =
        DescriptorSetLayoutBuilder::new(&mut pipeline_layout_builder);

    // TODO add these
    // default_descriptor_set_layout_builder.add_global_scope_parameters(program_layout);
    // default_descriptor_set_layout_builder.add_entry_point_parameters(program_layout);

    default_descriptor_set_layout_builder.build_and_add(&mut pipeline_layout_builder)?;

    pipeline_layout_builder.build()
}
