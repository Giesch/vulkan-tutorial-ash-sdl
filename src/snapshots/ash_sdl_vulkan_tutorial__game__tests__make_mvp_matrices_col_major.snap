---
source: src/game.rs
expression: mvp
---
{
  "model": [
    0.9848077,
    0.17364818,
    0.0,
    0.0,
    -0.17364818,
    0.9848077,
    0.0,
    0.0,
    0.0,
    0.0,
    1.0,
    0.0,
    0.0,
    0.0,
    0.0,
    1.0
  ],
  "view": [
    -0.70710677,
    -0.40824828,
    0.57735026,
    0.0,
    0.70710677,
    -0.40824828,
    0.57735026,
    0.0,
    0.0,
    0.81649655,
    0.57735026,
    0.0,
    -0.0,
    -0.0,
    -3.4641016,
    1.0
  ],
  "projection": [
    1.81066,
    0.0,
    0.0,
    0.0,
    0.0,
    -2.4142134,
    0.0,
    0.0,
    0.0,
    0.0,
    -1.0101011,
    -1.0,
    0.0,
    0.0,
    -0.10101011,
    0.0
  ]
}
